go-controller/pkg/ovn/egressfirewall_dns.go

```diff
@@ -0,0 +1,220 @@
+
+		added:          make(chan string, 1),
+		stopChan:       make(chan struct{}),
+		controllerStop: controllerStop,
+	}
+
+	go egressDNS.Sync()
+
+	return egressDNS, nil
+}
+
+func (e *EgressDNS) Add(namespace, dnsName string) (AddressSet, error) {
+	e.lock.Lock()
+	defer e.lock.Unlock()
+
+	if _, exists := e.dnsEntries[dnsName]; !exists {
+		var err error
+		dnsEntry := dnsEntry{
+			namespaces: make(map[string]struct{}),
+		}
+		if e.addressSetFactory == nil {
+			panic("addressSetFactory is null, that cannot happen")
+		}
+		dnsEntry.dnsAddressSet, err = e.addressSetFactory.NewAddressSet(dnsName, nil)
+		if err != nil {
+			return nil, fmt.Errorf("cannot create addressSet for %s: %v", dnsName, err)
+		}
+		e.dnsEntries[dnsName] = &dnsEntry
+		e.signalAdded(dnsName)
+	}
+	e.dnsEntries[dnsName].namespaces[namespace] = struct{}{}
+	return e.dnsEntries[dnsName].dnsAddressSet, nil
+
+}
+
+func (e *EgressDNS) Delete(namespace string) bool {
+	e.lock.Lock()
+	defer e.lock.Unlock()
+
+	// go through all dnsNames for namespaces
+	for dnsName, dnsEntry := range e.dnsEntries {
+		// delete the dnsEntry
+		delete(dnsEntry.namespaces, namespace)
+		if len(dnsEntry.namespaces) == 0 {
+			// the dnsEntry appears in no other namespace so delete the address_set
+			err := dnsEntry.dnsAddressSet.Destroy()
+			if err != nil {
+				klog.Errorf("Error deleteing EgressFirewall AddressSet for dnsName: %s %v", dnsName, err)
+			}
+			// the dnsEntry is no longer needed because nothing referances it delete it
+			delete(e.dnsEntries, dnsName)
+		}
+	}
+	return len(e.dnsEntries) == 0
+}
+
+func (e *EgressDNS) Update(dns string) (bool, error) {
+	e.lock.Lock()
+	defer e.lock.Unlock()
+
+	return e.dns.Update(dns)
+}
+
+func (e *EgressDNS) updateEntryForName(dnsName string) error {
+	e.lock.Lock()
+	defer e.lock.Unlock()
+
+	ips := e.dns.GetIPs(dnsName)
+	e.dnsEntries[dnsName].dnsResolves = ips
+	if err := e.dnsEntries[dnsName].dnsAddressSet.SetIPs(ips); err != nil {
+		return fmt.Errorf("cannot add IPs from EgressFirewall AddressSet %s: %v", dnsName, err)
+	}
+	return nil
+}
+
+var SyncDuration = 30 * time.Minute
+
+// Sync is a goroutine that constantly syncs the dns rules. Every dns name has it's own addressset
+// that the dns query updates whenever it is run. The sync loop waits and runs on after receiving
+// one of two signals.
+// 1. a new dnsName has been added
+// 2. the minimum of either 30 minutes or the next query time of a dns name
+// if a shutdown signal is seen the goroutine exits
+func (e *EgressDNS) Sync() {
+	var duration time.Duration
+	var dnsName string
+	for {
+		duration, dnsName = e.GetNextQueryTime()
+		if duration == 0 {
+			if _, err := e.Update(dnsName); err != nil {
+				utilruntime.HandleError(err)
+			}
+			if err := e.updateEntryForName(dnsName); err != nil {
+				utilruntime.HandleError(err)
+			}
+			continue
+		}
+		// Wait for duration or we recive the added signal
+		select {
+		case <-e.controllerStop:
```

Comment:
@JacobTanenbaum I think controllerStop and stopChan could just be passed into Sync() and not stored in the EgressDNS struct, since it looks like this is the only place they are used. And if you do get rid of stopChan (because per my other comment we don't kill it when the CRD goes away) then all we need is controllerStop (which I'd then rename to stopChan too)